### DISCUSSION TITLE
Selection Event

### DISCUSSION THREAD

**Discussion by kyledecot** [CoderLine/alphaTab#732] [Ideas]
I don't believe there is currently a way tell if the user is selecting a subset of the rendered track without digging down into private information (`_selectionStart` / `_selectionEnd`). It would be useful to expose the selection (or lack thereof) via an event.

**Comment by Danielku15** [comment]
This information is exposed as [`playbackRange`](https://alphatab.net/docs/reference/api/playbackrange/) describing the time range which should be played. This comes from the fact that in cases where multiple voices or tracks are rendered, the selection range is not guaranteed to be on the same hierarchy within the [score model](https://alphatab.net/docs/reference/score). 

Unfortunately the `tickLookup` is not exposed which would allow optimized lookup of the beat related to a given tick so a mapping from tick to beat would need to be done manually.

**Reply by Danielku15** [reply]
Update: I quickly exposed it.

**Reply by kyledecot** [reply]
💯  Thanks for exposing that additional property @Danielku15 – I'm not sure it helps me though. To give a little insight I'm using React in conjunction w/ AlphaTab. I want to be able to watch/observe changes to `playbackRange` so that I can make some UI updates. 

Knowing that I don't have to dig into `_selectionStart` and `_selectionEnd` gets me 1/2 of the way there but AKAIK there isn't a good event to hook into to get updates when the user selects/deselects/changes selection. I was thinking of something like:

```
const handlePlaybackRangeChange = useCallback((playbackRange) => {
  if (playbackRange) {
    // ...
  } else {
    // ...
  }
}, []);

useEffect(() => {
  if (!at) {
    return;
  }

  at.playbackRangeChanged.on(handlePlaybackRangeChange);

  return () => {
    at.playbackRangeChanged.off(handlePlaybackRangeChange);
  };
}, [at, handlePlaybackRangeChange]);
```

**Reply by Danielku15** [reply]
Have you considered using [`beatMouseDown`](https://alphatab.net/docs/reference/api/beatmousedown/), [beatMouseMove](https://alphatab.net/docs/reference/api/beatmousemove/) and [`beatMouseUp`](https://alphatab.net/docs/reference/api/beatmouseup/)? 

Those events are emitted during selection and would even give you access to the beat. If you only need the final new playback range when the selection is complete `beatMouseUp` is sufficient, after that you can access the `playbackRange` property to get the exact time range of selection. 

This would only work in case of end-user interaction/selection. For any kind of programmatic update I just merged a new event [`playbackRangeChanged`](https://github.com/CoderLine/alphaTab/pull/782). Docs here: https://alphatab.net/docs/reference/api/playbackrangechanged/

**Reply by kyledecot** [reply]
I had looked at `beatMouseUp` however this didn't work for me as I was programmatically modifying the selection. Thanks for adding https://github.com/CoderLine/alphaTab/pull/782 💯
